Jib pull base image from public registry and push to private registry

**Describe the bug**
Building an image with the container-jib extension fails, if the base image is hosted in a public registry and the properties `quarkus.container-image.username` and `quarkus.container-image.password` are set to push on a different private registry.

This is a regression from https://github.com/quarkusio/quarkus/pull/8288.

**Expected behavior**
Build image from the public registry, and push the image to a different private registry.

**Actual behavior**
```
20695 [ERROR] Caused by: com.google.cloud.tools.jib.api.RegistryAuthenticationFailedException: Failed to authenticate with registry registry-1.docker.io/azul/zulu-openjdk because: 401 Unauthorized
20695 [ERROR] {"details":"incorrect username or password"}
20695 [ERROR] 
20695 [ERROR] 	at com.google.cloud.tools.jib.registry.RegistryAuthenticator.authenticate(RegistryAuthenticator.java:306)
20695 [ERROR] 	at com.google.cloud.tools.jib.registry.RegistryAuthenticator.authenticate(RegistryAuthenticator.java:258)
20696 [ERROR] 	at com.google.cloud.tools.jib.registry.RegistryAuthenticator.authenticatePull(RegistryAuthenticator.java:177)
20696 [ERROR] 	at com.google.cloud.tools.jib.registry.RegistryClient.doBearerAuth(RegistryClient.java:362)
20696 [ERROR] 	at com.google.cloud.tools.jib.registry.RegistryClient.doPullBearerAuth(RegistryClient.java:327)
20696 [ERROR] 	at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call(PullBaseImageStep.java:158)
20696 [ERROR] 	at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call(PullBaseImageStep.java:56)
20697 [ERROR] 	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
20697 [ERROR] 	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
20697 [ERROR] 	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
20697 [ERROR] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
20697 [ERROR] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
20698 [ERROR] 	at java.base/java.lang.Thread.run(Thread.java:832)
```

**To Reproduce**
Steps to reproduce the behavior:
1. Set `quarkus.jib.base-jvm-image=docker.io/azul/zulu-openjdk:14`
2. Set `quarkus.container-image.registry=registry.gitlab.com`, `quarkus.container-image.group` and `quarkus.container-image.name`.
3. `./mvnw package -Dquarkus.container-image.push=true -Dquarkus.container-image.username=REGISTRY_USER -Dquarkus.container-image.password=REGISTRY_PASSWORD`

**Configuration**
```properties
quarkus.jib.base-jvm-image=docker.io/azul/zulu-openjdk:14
quarkus.container-image.registry=registry.gitlab.com
quarkus.container-image.group=org/group
quarkus.container-image.name=project
```

**Environment (please complete the following information):**
 - Output of `java -version`: openjdk version "14" 2020-03-17
 - Quarkus version or git rev: 1.3.2.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Apache Maven 3.6.3

**Additional context**
The public base image is set to `docker.io` explicitly (it's the default if not set), the `quarkus.container-image.registry` is set to `registry.gitlab.com`, since both registries are different one approach is to detect if they are the same to authenticate using the container-image properties. 

Thinking more on this, if both registries are different (both private) the best would be to provide another set of permissions (user/pass) to be used by Jib to pull and push.

--- a/extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibConfig.java
+++ b/extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibConfig.java
@@ -2,6 +2,7 @@ package io.quarkus.container.image.jib.deployment;
 
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 
 import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigPhase;
@@ -44,5 +45,17 @@ public class JibConfig {
      * Custom labels to add to the generated image
      */
     @ConfigItem
-    Map<String, String> labels;
+    public Map<String, String> labels;
+
+    /**
+     * The username to use to authenticate with the registry used to pull the base JVM image
+     */
+    @ConfigItem
+    public Optional<String> baseRegistryUsername;
+
+    /**
+     * The password to use to authenticate with the registry used to pull the base JVM image
+     */
+    @ConfigItem
+    public Optional<String> baseRegistryPassword;
 }

--- a/extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java
+++ b/extensions/container-image/container-image-jib/deployment/src/main/java/io/quarkus/container/image/jib/deployment/JibProcessor.java
@@ -74,9 +74,9 @@ public class JibProcessor {
             return;
         }
 
-        JibContainerBuilder jibContainerBuilder = createContainerBuilderFromJar(containerImageConfig, jibConfig,
+        JibContainerBuilder jibContainerBuilder = createContainerBuilderFromJar(jibConfig,
                 sourceJar, outputTarget, mainClass, containerImageLabels);
-        JibContainer container = containerize(applicationInfo, containerImageConfig, jibConfig, jibContainerBuilder,
+        JibContainer container = containerize(applicationInfo, containerImageConfig, jibContainerBuilder,
                 pushRequest.isPresent());
 
         ImageReference targetImage = container.getTargetImage();
@@ -107,7 +107,7 @@ public class JibProcessor {
 
         JibContainerBuilder jibContainerBuilder = createContainerBuilderFromNative(containerImageConfig, jibConfig,
                 nativeImage, containerImageLabels);
-        JibContainer container = containerize(applicationInfo, containerImageConfig, jibConfig, jibContainerBuilder,
+        JibContainer container = containerize(applicationInfo, containerImageConfig, jibContainerBuilder,
                 pushRequest.isPresent());
 
         ImageReference targetImage = container.getTargetImage();
@@ -117,9 +117,8 @@ public class JibProcessor {
     }
 
     private JibContainer containerize(ApplicationInfoBuildItem applicationInfo, ContainerImageConfig containerImageConfig,
-            JibConfig jibConfig,
             JibContainerBuilder jibContainerBuilder, boolean pushRequested) {
-        Containerizer containerizer = createContainerizer(containerImageConfig, jibConfig, applicationInfo, pushRequested);
+        Containerizer containerizer = createContainerizer(containerImageConfig, applicationInfo, pushRequested);
         try {
             log.info("Starting container image build");
             JibContainer container = jibContainerBuilder.containerize(containerizer);
@@ -133,7 +132,7 @@ public class JibProcessor {
         }
     }
 
-    private Containerizer createContainerizer(ContainerImageConfig containerImageConfig, JibConfig jibConfig,
+    private Containerizer createContainerizer(ContainerImageConfig containerImageConfig,
             ApplicationInfoBuildItem applicationInfo, boolean pushRequested) {
         Containerizer containerizer;
         ImageReference imageReference = getImageReference(containerImageConfig, applicationInfo);
@@ -142,7 +141,8 @@ public class JibProcessor {
             if (!containerImageConfig.registry.isPresent()) {
                 log.info("No container image registry was set, so 'docker.io' will be used");
             }
-            RegistryImage registryImage = toRegistryImage(imageReference, containerImageConfig);
+            RegistryImage registryImage = toRegistryImage(imageReference, containerImageConfig.username,
+                    containerImageConfig.password);
             containerizer = Containerizer.to(registryImage);
         } else {
             containerizer = Containerizer.to(DockerDaemonImage.named(imageReference));
@@ -157,14 +157,14 @@ public class JibProcessor {
         return containerizer;
     }
 
-    private RegistryImage toRegistryImage(ImageReference imageReference, ContainerImageConfig containerImageConfig) {
+    private RegistryImage toRegistryImage(ImageReference imageReference, Optional<String> username, Optional<String> password) {
         CredentialRetrieverFactory credentialRetrieverFactory = CredentialRetrieverFactory.forImage(imageReference,
                 log::info);
         RegistryImage registryImage = RegistryImage.named(imageReference);
         registryImage.addCredentialRetriever(credentialRetrieverFactory.wellKnownCredentialHelpers());
         registryImage.addCredentialRetriever(credentialRetrieverFactory.dockerConfig());
-        if (containerImageConfig.username.isPresent() && containerImageConfig.password.isPresent()) {
-            registryImage.addCredential(containerImageConfig.username.get(), containerImageConfig.password.get());
+        if (username.isPresent() && password.isPresent()) {
+            registryImage.addCredential(username.get(), password.get());
         }
         return registryImage;
     }
@@ -190,7 +190,7 @@ public class JibProcessor {
                 containerImageConfig.tag.orElse(applicationInfo.getVersion()));
     }
 
-    private JibContainerBuilder createContainerBuilderFromJar(ContainerImageConfig containerImageConfig, JibConfig jibConfig,
+    private JibContainerBuilder createContainerBuilderFromJar(JibConfig jibConfig,
             JarBuildItem sourceJarBuildItem,
             OutputTargetBuildItem outputTargetBuildItem, MainClassBuildItem mainClassBuildItem,
             List<ContainerImageLabelBuildItem> containerImageLabels) {
@@ -199,7 +199,8 @@ public class JibProcessor {
             Path classesDir = outputTargetBuildItem.getOutputDirectory().resolve("jib");
             ZipUtils.unzip(sourceJarBuildItem.getPath(), classesDir);
             JavaContainerBuilder javaContainerBuilder = JavaContainerBuilder
-                    .from(toRegistryImage(ImageReference.parse(jibConfig.baseJvmImage), containerImageConfig))
+                    .from(toRegistryImage(ImageReference.parse(jibConfig.baseJvmImage), jibConfig.baseRegistryUsername,
+                            jibConfig.baseRegistryPassword))
                     .addResources(classesDir, IS_CLASS_PREDICATE.negate())
                     .addClasses(classesDir, IS_CLASS_PREDICATE)
                     .addJvmFlags(jibConfig.jvmArguments)
@@ -231,7 +232,9 @@ public class JibProcessor {
         entrypoint.addAll(jibConfig.nativeArguments);
         try {
             AbsoluteUnixPath workDirInContainer = AbsoluteUnixPath.get("/work");
-            return Jib.from(toRegistryImage(ImageReference.parse(jibConfig.baseNativeImage), containerImageConfig))
+            return Jib
+                    .from(toRegistryImage(ImageReference.parse(jibConfig.baseNativeImage), containerImageConfig.username,
+                            containerImageConfig.password))
                     .addLayer(LayerConfiguration.builder()
                             .addEntry(nativeImageBuildItem.getPath(), workDirInContainer.resolve(BINARY_NAME_IN_CONTAINER),
                                     FilePermissions.fromOctalString("775"))

--- a/extensions/container-image/deployment/src/main/java/io/quarkus/container/image/deployment/ContainerImageConfig.java
+++ b/extensions/container-image/deployment/src/main/java/io/quarkus/container/image/deployment/ContainerImageConfig.java
@@ -33,13 +33,13 @@ public class ContainerImageConfig {
     public Optional<String> registry;
 
     /**
-     * The username to use to authenticate with the registry
+     * The username to use to authenticate with the registry where the built image will be pushed
      */
     @ConfigItem
     public Optional<String> username;
 
     /**
-     * The password to use to authenticate with the registry
+     * The password to use to authenticate with the registry where the built image will be pushed
      */
     @ConfigItem
     public Optional<String> password;

